README.md
## elf-tools

A NodeJS library for building and parsing ELF (Executable and Linker Format) binaries.

### Building an ELF image

There are two ways to output an ELF image:

#### build(opts)
Use `elf_tools.build(opts)` to create a buffer containing ELF bytes
```javascript
const fs = require('fs');
const elf_tools = require('elf-tools');

// create some raw "Hello world!" code for the ELF executable
// note that these instruction bytes are for the Linux OS on Intel/AMD x64 platform
const code = Buffer.from(
 '488d3523000000'    // lea    rsi,[rip+0x23]
+'48c7c20d000000'    // mov    rdx,0xd
+'48c7c701000000'    // mov    rdi,0x1
+'48c7c001000000'    // mov    rax,0x1
+'0f05'              // syscall
+'4831ff'            // xor    rdi,rdi
+'48c7c03c000000'    // mov    rax,0x3c
+'0f05'              // syscall
+'48656c6c6f20776f726c64210a'	// "Hello world!\n"
, 'hex');

// build the ELF executable with the code
const image = elf_tools.build({
    code,
});

// save the ELF image as a binary executable file
fs.writeFileSync('./a.out', image, { mode: 0o755 });

```

#### createBuildStream(opts)
Use `elf_tools.createBuildStream(opts)` and pipe the result to another [Writable](https://nodejs.org/api/stream.html#stream_writable_streams) stream
```javascript
const fs = require('fs');
const elf_tools = require('elf-tools');

// create some raw "Hello world!" code for the ELF executable
// note that these instruction bytes are for the Linux OS on Intel/AMD x64 platform
const code = Buffer.from(
 '488d3523000000'    // lea    rsi,[rip+0x23]
+'48c7c20d000000'    // mov    rdx,0xd
+'48c7c701000000'    // mov    rdi,0x1
+'48c7c001000000'    // mov    rax,0x1
+'0f05'              // syscall
+'4831ff'            // xor    rdi,rdi
+'48c7c03c000000'    // mov    rax,0x3c
+'0f05'              // syscall
+'48656c6c6f20776f726c64210a'	// "Hello world!\n"
, 'hex');

// create a ELF stream to output the ELF image bytes
const elf_stream = elf_tools.createBuildStream({
    code,
});

// create a writable file stream
const file_stream = fs.createWriteStream('./a.out', { mode: 0o755 });

// pipe the ELF bytes to the file
elf_stream.pipe(file_stream);

```
Once the file is saved, run it from a terminal:
```
$ ./a.out
Hello world!
```

### Parsing ELF files

To parse an ELF file:
```javascript
const fs = require('fs');
const elf_tools = require('elf-tools');

// read the file into a buffer
const elf_bytes = fs.readFileSync('./a.out');

// parse the data
const elf = elf_tools.parse(elf_bytes);

// elf is an object containing the parsed ELF-header, program headers (if any) and sections

```

### Documentation

### `build(opts)`

Build an ELF image
- **opts**: Object containing code and data to place inside the ELF image
  - code: **Buffer**. Raw executable code bytes. The ELF builder places this data in a *.text* section with the ELF entry point pointing to the first byte. All calls to `build()` must include a code buffer. The caller must ensure the code bytes represent valid instructions for the target machine. **Important**: the code must be position-independant with no relocations.
  - rodata: **Buffer** (optional). Raw read-only data bytes. The ELF builder places any read-only data in the *.text* section immediately following the code buffer. If this data needs aligning to a particular byte boundary, the code buffer should be padded accordingly.
  - rwdata: **Buffer** (optional). Raw read/write data bytes. The ELF builder places any writable data in a *.data* section following the code and read-only data
  - bss_length: **integer** (optional). Length of uninitialised (writable) data bytes. The ELF builder locates this data after the rwdata buffer. Although the data is "uninitialized" (because it takes up no space in the ELF image), when the image is run, this section is always zero-filled.
  
- returns **Buffer** containing the complete ELF image

#### Example

This example uses all 4 fields (code, r/o data, r/w data, bss data) to build an executable which:
- writes out a prompt (r/o data)
- then reads user-inputted text into a buffer (r/w data)
- then reverses the text using scratch space (bss data)
- finally writes out the reversed text and exits the program

```javascript
const fs = require('fs');
const elf_tools = require('elf-tools');

const opts = {
    // note that these instruction bytes are for the Linux OS on Intel/AMD x64 platform
    // you can find the source to this program at https://github.com/adelphes/elf-tools/blob/master/test/programs/reverse/reverse.s
    code: Buffer.from(`
488d3579 00000048 c7c21c00 0000e85c 00000048 8d358600 000048c7 c2000100
00e83e00 000048ff c84889c2 488d3d6d 01000048 01c78a0e 880f48ff cf48ffc6
4883e801 75f04889 fe48ffc6 c604160a 48ffc2e8 17000000 4831ff48 c7c03c00
00000f05 bf010000 004831c0 0f05c348 c7c70100 000048c7 c0010000 000f05c3
    `.replace(/\s+/g,''), 'hex'),
    
    // add the prompt as read-only bytes - the extra null bytes pad the section to an 8-byte alignment.
    rodata: Buffer.from('Enter some text to reverse: \0\0\0\0', 'ascii'),

    // pre-allocate 256 zero bytes in the program to store the user-input
    rwdata: Buffer.alloc(256),

    // reserve another 256 bytes of memory for us to reverse the text
    bss_length: 256,
}

// create a ELF stream to output the ELF image bytes
const elf_stream = elf_tools.createStream(opts);

// create a writable file stream
const file_stream = fs.createWriteStream('./reverse', { mode: 0o755 });

// pipe the ELF bytes to the file
elf_stream.pipe(file_stream);

```
Once the file is saved, run it
```
$ ./reverse
Enter some text to reverse: abcdef
fedcba
```


CHANGELOG.md
## 1.1.0
- Allow image load address, entrypoint and ELF headers to be customised
- Set default ELF header values based on executing platform

## 1.0.0
- Initial release


test/test-reader.js
const assert = require('assert');
const elf_tools = require('../index.js');
const fs = require('fs');
const path = require('path');

describe("ELF reader", function () {
    it("parses valid ELF files", function () {
        const binary = fs.readFileSync(path.join(__dirname, 'programs/helloworld/helloworld'));
        assert.doesNotThrow(() => {
            const parsed = elf_tools.parse(binary);
            assert.ok(parsed);
        });
    });


    it("throws on bad ELF signature", function () {
        const binary = Buffer.from('BAD_ELF_SIG');
        assert.throws(() => {
            elf_tools.parse(binary);
        });
    });
});


test/test-writer.js
const assert = require('assert');
const elf_tools = require('../index.js');
const fs = require('fs');
const path = require('path');
const os = require('os');
const childproc = require('child_process');

function make_byte_buffer(str) {
    // strip any whitespace and comments
    const hex = str.replace(/(\s+)|(#.*)/g,'');
    return Buffer.from(hex, 'hex');
}

describe("ELF writer", function () {

    it("builds with code Buffer parameter", function () {
        const code = make_byte_buffer(`
            48 31 ff                # xor    rdi,rdi
            48 c7 c0 3c 00 00 00    # mov    rax,0x3c
            0f 05                   # syscall
        `);
        const elf_image = elf_tools.build(code);
        const expected = make_byte_buffer(`7f454c4602010100000000000000000002003e0001000000b0004000000000004000000000000000d800000000000000000000004000380002004000040003000
            100000005000000000000000000000000004000000000000000400000000000bc00000000000000bc0000000000000000001000000000000100000006000000bc00000000000000bc00400000000000bc004
            000000000000000000000000000000000000000000008000000000000004831ff48c7c03c0000000f05002e74657874002e64617461002e73687374727461620000000000000000000000000000000000000
            000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000010000000600000000000000b000400000000000b000000000000
            0000c0000000000000000000000000000000100000000000000000000000000000007000000010000000300000000000000bc00400000000000bc00000000000000000000000000000000000000000000000
            10000000000000000000000000000000d0000000300000000000000000000000000000000000000bc000000000000001700000000000000000000000000000001000000000000000000000000000000`);
        assert.ok(elf_image.equals(expected));
    });

    it("streams to a writable file", function () {
        const code = make_byte_buffer(`
            48 31 ff                # xor    rdi,rdi
            48 c7 c0 3c 00 00 00    # mov    rax,0x3c
            0f 05                   # syscall
        `);
        const elf_stream = elf_tools.createBuildStream(code);
        const tmpfile = path.join(os.tmpdir(), `elftest-${process.hrtime()[1]}` );
        const file_stream = fs.createWriteStream(tmpfile, {mode: 0o755});
        return new Promise((res, rej) => {
            file_stream.on('error', rej);
            file_stream.on('close', () => {
                fs.unlinkSync(tmpfile);
                res();
            })
            elf_stream.pipe(file_stream);
        });
    });

    it("builds an executable that will run", function () {
        if (process.arch !== 'x64') {
            this.skip();
            return;
        }
        const code = make_byte_buffer(`
            48 31 ff                # xor    rdi,rdi
            48 c7 c0 3c 00 00 00    # mov    rax,0x3c
            0f 05                   # syscall
        `);
        const elf_image = elf_tools.build(code);
        const tmpfile = path.join(os.tmpdir(), `elftest-${process.hrtime()[1]}` );
        fs.writeFileSync(tmpfile, elf_image, {mode: 0o755});
        try {
            const result = childproc.execFileSync(tmpfile);
            assert.ok(result);
        } finally {
            fs.unlinkSync(tmpfile);
        }
    });

    it("builds with object parameter with single code Buffer field", function () {
        const code = make_byte_buffer(`
            48 31 ff                # xor    rdi,rdi
            48 c7 c0 3c 00 00 00    # mov    rax,0x3c
            0f 05                   # syscall
        `);
        elf_tools.build({
            code,
        });
    });

    it("allows a custom image load address", function () {
        const code = make_byte_buffer(`
            48 31 ff                # xor    rdi,rdi
            48 c7 c0 3c 00 00 00    # mov    rax,0x3c
            0f 05                   # syscall
        `);
        const base_address = 0x12340000;
        const elf_image = elf_tools.build({
            code,
            base_address,
        });
        const entry_point = elf_image.readUInt32LE(0x18);
        assert.strictEqual(entry_point & 0xffff0000, 0x12340000);
    });

    it("allows a custom entry point offset", function () {
        const code = make_byte_buffer(`
            90                      # nop
            48 31 ff                # xor    rdi,rdi
            48 c7 c0 3c 00 00 00    # mov    rax,0x3c
            0f 05                   # syscall
        `);
        const base_address = 0x12340000;
        const entry_offset = 1; // skip the nop
        const elf_image = elf_tools.build({
            code,
            base_address,
            entry_offset,
        });
        const entry_point = elf_image.readUInt32LE(0x18);
        assert.strictEqual(entry_point, base_address + 0xb0 + entry_offset);
    });

    it("allows custom ELF header values", function () {
        const code = make_byte_buffer(`
            48 31 ff                # xor    rdi,rdi
            48 c7 c0 3c 00 00 00    # mov    rax,0x3c
            0f 05                   # syscall
        `);
        const elf_header = {
            class: '32',
            endian: 'msb',
            osabi: 'arm',
            abiversion: 'none',
            type: 'exec',
            machine: 'arm',
        }
        elf_tools.build({
            code,
            elf_header,
        });
    });

    it("throws on missing parameter", function () {
        assert.throws(() => elf_tools.build());
    });

    it("throws on invalid parameter", function () {
        assert.throws(() => elf_tools.build(''));
        assert.throws(() => elf_tools.build(3));
        assert.throws(() => elf_tools.build(true));
        assert.throws(() => elf_tools.build(null));
    });

    it("throws on missing code field", function () {
        assert.throws(() => {
            elf_tools.build({});
        });
    });

    it("throws on zero-length code", function () {
        const code = Buffer.allocUnsafe(0);
        assert.throws(() => elf_tools.build(code));
        assert.throws(() => elf_tools.build({
            code
        }));
    });

    it("throws when base_address/entry_offset and elf_header.entry are defined", function () {
        const code = Buffer.allocUnsafe(1);
        assert.throws(() => elf_tools.build({
            code,
            base_address: 0x100000,
            elf_header: {
                entry: 0x200000,
            }
        }));
        assert.throws(() => elf_tools.build({
            code,
            entry_offset: 1,
            elf_header: {
                entry: 0x200000,
            }
        }));
    });
});


test/programs/reverse/reverse.s
.intel_syntax noprefix

# simple program to output a reversed string

# GNU assemble and link:
# as --64 -o ./reverse.o ./reverse.s && ld -s -Ttext 0x4000b0 -Tdata 0x400150 -o reverse ./reverse.o
# run:
# ./a.out

.globl _start

.text
_start:
  # write the prompt
  lea rsi, [rip + prompt]
  mov rdx, 28   # length of prompt
  call _write

  # read the user input
  lea rsi, [rip + buf]  # buffer
  mov rdx, 256  # max size
  call _read
  dec rax       # ignore newline
  mov rdx, rax  # save length

  # copy the string in reverse to the scratch area
  lea rdi, [rip + scratch]
  add rdi, rax
1:
  mov cl, [rsi]
  mov [rdi], cl
  dec rdi
  inc rsi
  sub rax, 1
  jnz 1b    # 1b(efore)

  # write out the reversed string
  mov rsi, rdi
  inc rsi
  mov [rsi+rdx], byte ptr 10    # append a newline to the end
  inc rdx
  call _write

  # exit process with code=0
  xor rdi, rdi  
  mov rax, 60	# sys_exit
  syscall

# int read(fd=rdi, buf=rsi, count=rdx);
_read:
  mov edi, 1	# fd=stdout
  xor rax, rax  # syscall=read
  syscall
  ret

# int write(fd=rdi, buf=rsi, count=rdx);
_write:
  mov rdi, 1	# fd=stdout
  mov rax, 1	# syscall=write
  syscall
  ret

prompt:
.ascii "Enter some text to reverse: "


.data 
# the .data section
.align 8
buf:
.space 256, 0


.bss
# the .bss section
scratch:
.space 256, 0


lib/elf-reader.js
const constants = require('./constants');
const ELFHeader = require('./elf_header');
const ProgramHeader = require('./program_header');
const SectionHeader = require('./section_header');
const StringTable = require('./string_table');

/**
 * Returns the integer-writing method that should be used
 * @param {'lsb'|'msb'} endian
 */
function getReadIntFn(endian) {
  switch (endian) {
    case 'lsb':
      return 'readUIntLE';
    case 'msb':
      return 'readUIntBE';
    default:
      new Error('Invalid endian: ' + endian);
  }
}

function parse(image) {
    // load the first part of the ELF-header with the signature, word-size and endian
  const hdr_bytes = image.slice(0, 8);
  const hdr = {
    signature: hdr_bytes.slice(0, 4),
    class: constants.class[hdr_bytes[4]],
    endian: constants.endian[hdr_bytes[5]],
    osabi: constants.osabi[hdr_bytes[6]],
    abiversion: constants.abiversion[hdr_bytes[7]],
  };
  if (!hdr.signature.equals(constants.ELF_SIGNATURE_BYTES)) {
    throw new Error('Invalid ELF signature bytes');
  }

  const stream = {
    image,
    idx: 0,
    word_size: hdr.class / 8,
    readIntFn: getReadIntFn(hdr.endian),
    seek(offset) {
        return this.idx = offset;
    },
    skip(sz) {
        return this.idx += sz;
    },
    readWord(sz) {
      sz = sz || this.word_size;
      this.idx += sz;
      return this.image[this.readIntFn](this.idx - sz, sz);
    },
    readBuf(sz) {
      return this.image.slice(this.idx, (this.idx += sz));
    },
  };

  // parse the header (fully)
  const header = ELFHeader.parse(stream);
  const programs = [];
  const sections = [];

  // parse the program headers
  stream.seek(header.phoff);
  for (let i=0; i < header.phnum; i++) {
    const ph = ProgramHeader.parse(stream, header);
    programs.push({
        header: ph,
        data: image.slice(ph.offset, ph.offset + ph.filesz),
      })
  }

  // parse the section headers
  stream.seek(header.shoff);
  for (let i=0; i < header.shnum; i++) {
    const sh = SectionHeader.parse(stream);
    sections.push({
        header: sh,
        data: image.slice(sh.offset, sh.offset + sh.size),
      })
  }

  // resolve the section header string table and convert the section header name offsets to real names
  const sh_string_table = StringTable.parse(sections[header.shstrndx]);
  sections.forEach(section => {
      section.header.name = sh_string_table.getString(section.header.name);
  });


  return {
      header,
      programs,
      sections,
      sh_string_table,
  }
}

module.exports = {
    parse,
}


lib/elf_header.js
const constants = require('./constants');

/**
 * ELFHeader is the main program header containing the ELF signature and
 * general machine, program and section information.
 */
class ELFHeader {
  constructor(header) {
    Object.assign(
      this,
      {
        elfsig: constants.ELF_SIGNATURE_BYTES,
        class: '64',
        endian: 'lsb',
        osabi: 'unix-sysv',
        abiversion: 'none',
        type: 'exec',
        machine: 'amd64',
        version: 1,
        entry: null,
        phoff: null,
        shoff: null,
        flags: 0,
        ehsize: null,
        phentsize: null,
        phnum: null,
        shentsize: null,
        shnum: null,
        shstrndx: null,
      },
      // override the default values with any custom values
      header
    );
  }

  static parse(stream) {
    return new ELFHeader({
      elfsig: stream.readBuf(4),
      class: constants.class[stream.readWord(1)],
      endian: constants.endian[stream.readWord(1)],
      osabi: constants.osabi[stream.readWord(1)],
      abiversion: constants.abiversion[stream.readWord(1)],
      padding: stream.readBuf(8),
      type: constants.type[stream.readWord(2)],
      machine: constants.machine[stream.readWord(2)],
      version: stream.readWord(4),
      entry: stream.readWord(),
      phoff: stream.readWord(),
      shoff: stream.readWord(),
      flags: stream.readWord(4),
      ehsize: stream.readWord(2),
      phentsize: stream.readWord(2),
      phnum: stream.readWord(2),
      shentsize: stream.readWord(2),
      shnum: stream.readWord(2),
      shstrndx: stream.readWord(2),
    });
  }

  calculate_size(elf_offset, elf) {
    this.elf_offset = elf_offset; // should always be 0
    this.elf_size = this.ehsize = elf.word_size * 3 + 40;
    return this.elf_size;
  }

  finalise(entrypoint, program_headers, section_headers, string_table_idx) {
    this.entry = entrypoint;
    this.phoff = program_headers[0].elf_offset;
    this.phentsize = program_headers[0].elf_size;
    this.phnum = program_headers.length;
    this.shoff = section_headers[0].elf_offset;
    this.shentsize = section_headers[0].elf_size;
    this.shnum = section_headers.length;
    this.shstrndx = string_table_idx;
  }

  write(stream) {
    stream.writeBuf(this.elfsig);
    stream.writeWord(constants.class[this.class], 1);
    stream.writeWord(constants.endian[this.endian], 1);
    stream.writeWord(constants.osabi[this.osabi], 1);
    stream.writeWord(constants.abiversion[this.abiversion], 1);
    stream.skip(8);
    stream.writeWord(constants.type[this.type], 2);
    stream.writeWord(constants.machine[this.machine], 2);
    stream.writeWord(this.version, 4);
    stream.writeWord(this.entry);
    stream.writeWord(this.phoff);
    stream.writeWord(this.shoff);
    stream.writeWord(this.flags, 4);
    stream.writeWord(this.ehsize, 2);
    stream.writeWord(this.phentsize, 2);
    stream.writeWord(this.phnum, 2);
    stream.writeWord(this.shentsize, 2);
    stream.writeWord(this.shnum, 2);
    stream.writeWord(this.shstrndx, 2);
  }
}

module.exports = ELFHeader;


lib/constants.js
const constants = {};

constants.ELF_SIGNATURE_BYTES = Buffer.from('\x7fELF', 'ascii');

constants.class = {
  0: 'none',
  1: '32',
  2: '64',
  none: 0,
  '32': 1,
  '64': 2,
};

constants.endian = {
  0: 'none',
  1: 'lsb',
  2: 'msb',
  none: 0,
  lsb: 1,
  msb: 2,
};

constants.version = {
  0: 'none',
  1: 'current',
  none: 0,
  current: 1,
};

constants.osabi = {
  0: 'sysv',
  1: 'unix-sysv',
  2: 'netbsd',
  3: 'linux',
  6: 'solaris',
  7: 'aix',
  8: 'irix',
  9: 'freebsd',
  10: 'tru64',
  11: 'modesto',
  12: 'openbsd',
  13: 'openvms',
  14: 'nsk',
  15: 'aros',
  97: 'arm',
  255: 'standalone',
  sysv: 0,
  'unix-sysv': 1,
  netbsd: 2,
  linux: 3,
  solaris: 6,
  aix: 7,
  irix: 8,
  freebsd: 9,
  tru64: 10,
  modesto: 11,
  openbsd: 12,
  openvms: 13,
  nsk: 14,
  aros: 15,
  arm: 97,
  standalone: 255,
};

constants.abiversion = {
  0: 'none',
  1: 'current',
  none: 0,
  current: 1,
};

constants.machine = {
  0: 'none',
  1: 'm32',
  2: 'sparc',
  3: '386',
  4: '68k',
  5: '88k',
  6: '486',
  7: '860',
  8: 'mips',
  9: 's370',
  10: 'mips_rs3_le',
  11: 'rs6000',
  15: 'pa_risc',
  16: 'ncube',
  17: 'vpp500',
  18: 'sparc32plus',
  19: '960',
  20: 'ppc',
  21: 'ppc64',
  22: 's390',
  36: 'v800',
  37: 'fr20',
  38: 'rh32',
  39: 'rce',
  40: 'arm',
  41: 'alpha',
  42: 'sh',
  43: 'sparcv9',
  44: 'tricore',
  45: 'arc',
  46: 'h8_300',
  47: 'h8_300h',
  48: 'h8s',
  49: 'h8_500',
  50: 'ia_64',
  51: 'mips_x',
  52: 'coldfire',
  53: '68hc12',
  54: 'mma',
  55: 'pcp',
  56: 'ncpu',
  57: 'ndr1',
  58: 'starcore',
  59: 'me16',
  60: 'st100',
  61: 'tinyj',
  62: 'amd64',
  63: 'pdsp',
  66: 'fx66',
  67: 'st9plus',
  68: 'st7',
  69: '68hc16',
  70: '68hc11',
  71: '68hc08',
  72: '68hc05',
  73: 'svx',
  74: 'st19',
  75: 'vax',
  76: 'cris',
  77: 'javelin',
  78: 'firepath',
  79: 'zsp',
  80: 'mmix',
  81: 'huany',
  82: 'prism',
  83: 'avr',
  84: 'fr30',
  85: 'd10v',
  86: 'd30v',
  87: 'v850',
  88: 'm32r',
  89: 'mn10300',
  90: 'mn10200',
  91: 'pj',
  92: 'openrisc',
  93: 'arc_a5',
  94: 'xtensa',
  none: 0,
  m32: 1,
  sparc: 2,
  '68k': 4,
  '88k': 5,
  mips: 8,
  s370: 9,
  mips_rs3_le: 10,
  rs6000: 11,
  pa_risc: 15,
  ncube: 16,
  vpp500: 17,
  sparc32plus: 18,
  ppc: 20,
  ppc64: 21,
  s390: 22,
  v800: 36,
  fr20: 37,
  rh32: 38,
  rce: 39,
  arm: 40,
  alpha: 41,
  sh: 42,
  sparcv9: 43,
  tricore: 44,
  arc: 45,
  h8_300: 46,
  h8_300h: 47,
  h8s: 48,
  h8_500: 49,
  ia_64: 50,
  mips_x: 51,
  coldfire: 52,
  '68hc12': 53,
  mma: 54,
  pcp: 55,
  ncpu: 56,
  ndr1: 57,
  starcore: 58,
  me16: 59,
  st100: 60,
  tinyj: 61,
  amd64: 62,
  pdsp: 63,
  fx66: 66,
  st9plus: 67,
  st7: 68,
  '68hc16': 69,
  '68hc11': 70,
  '68hc08': 71,
  '68hc05': 72,
  svx: 73,
  st19: 74,
  vax: 75,
  cris: 76,
  javelin: 77,
  firepath: 78,
  zsp: 79,
  mmix: 80,
  huany: 81,
  prism: 82,
  avr: 83,
  fr30: 84,
  d10v: 85,
  d30v: 86,
  v850: 87,
  m32r: 88,
  mn10300: 89,
  mn10200: 90,
  pj: 91,
  openrisc: 92,
  arc_a5: 93,
  xtensa: 94,
};

constants.type = {
  '0': 'none',
  '1': 'rel',
  '2': 'exec',
  '3': 'dyn',
  '4': 'core',
  none: 0,
  rel: 1,
  exec: 2,
  dyn: 3,
  core: 4,
};

constants.entryType = {
  0: 'null',
  1: 'load',
  2: 'dynamic',
  3: 'interp',
  4: 'note',
  5: 'shlib',
  6: 'phdr',
  7: 'tls',
  null: 0,
  load: 1,
  dynamic: 2,
  interp: 3,
  note: 4,
  shlib: 5,
  phdr: 6,
  tls: 7,
  0x6464e550: 'sunw_unwind',
  0x6474e550: 'sunw_eh_frame',
  0x6474e551: 'gnu_stack',
  0x6474e552: 'gnu_relro',
  0x6ffffffa: 'losunw',
  0x6ffffffa: 'sunwbss',
  0x6ffffffb: 'sunwstack',
  0x6ffffffc: 'sunwdtrace',
  0x6ffffffd: 'sunwcap',
};

constants.entryFlags = {
  4: 'read',
  2: 'write',
  1: 'execute',
  read: 4,
  write: 2,
  execute: 1,
  0x00100000: 'sunw_failure',
  0x00200000: 'sunw_killed',
  0x00400000: 'sunw_siginfo',
};

constants.sectType = {
  0: 'null',
  1: 'progbits',
  2: 'symtab',
  3: 'strtab',
  4: 'rela',
  5: 'hash',
  6: 'dynamic',
  7: 'note',
  8: 'nobits',
  9: 'rel',
  10: 'shlib',
  11: 'dynsym',
  12: 'unknown12',
  13: 'unknown13',
  14: 'init_array',
  15: 'fini_array',
  16: 'preinit_array',
  17: 'group',
  18: 'symtab_shndx',

  null: 0,
  progbits: 1,
  symtab: 2,
  strtab: 3,
  rela: 4,
  hash: 5,
  dynamic: 6,
  note: 7,
  nobits: 8,
  rel: 9,
  shlib: 10,
  dynsym: 11,
  unknown12: 12,
  unknown13: 13,
  init_array: 14,
  fini_array: 15,
  preinit_array: 16,
  group: 17,
  symtab_shndx: 18,

  0x60000000: 'loos',
  0x6fffffef: 'sunw_capchain',
  0x6ffffff0: 'sunw_capinfo',
  0x6ffffff1: 'sunw_symsort',
  0x6ffffff2: 'sunw_tlssort',
  0x6ffffff3: 'sunw_ldynsym',
  0x6ffffff4: 'sunw_dof',
  0x6ffffff5: 'sunw_cap',
  0x6ffffff6: 'sunw_signature',
  0x6ffffff7: 'sunw_annotate',
  0x6ffffff8: 'sunw_debugstr',
  0x6ffffff9: 'sunw_debug',
  0x6ffffffa: 'sunw_move',
  0x6ffffffb: 'sunw_comdat',
  0x6ffffffc: 'sunw_syminfo',
  0x6ffffffd: 'sunw_verdef',
  0x6ffffffe: 'sunw_verneed',
  0x6fffffff: 'sunw_versym',
  0x6fffffff: 'hisunw',
  0x6fffffff: 'hios',
  0x6ffffff5: 'gnu_attributes',
  0x6ffffff6: 'gnu_hash',
  0x6ffffff7: 'gnu_liblist',
  0x6ffffff8: 'checksum',
};

constants.sectFlags = {
  0x01: 'write',
  0x02: 'alloc',
  0x04: 'execute',
  0x10: 'merge',
  0x20: 'strings',
  0x40: 'info_link',
  0x80: 'link_order',
  0x100: 'os_nonconforming',
  0x200: 'group',
  0x400: 'tls',
  write: 0x1,
  alloc: 0x2,
  execute: 0x4,
  merge: 0x10,
  strings: 0x20,
  info_link: 0x40,
  link_order: 0x80,
  os_nonconforming: 0x100,
  group: 0x200,
  tls: 0x400,
};

constants.dynTypes = {
  '0': 'DT_NULL',
  '1': 'DT_NEEDED',
  '2': 'DT_PLTRELSZ',
  '3': 'DT_PLTGOT',
  '4': 'DT_HASH',
  '5': 'DT_STRTAB',
  '6': 'DT_SYMTAB',
  '7': 'DT_RELA',
  '8': 'DT_RELASZ',
  '9': 'DT_RELAENT',
  '10': 'DT_STRSZ',
  '11': 'DT_SYMENT',
  '12': 'DT_INIT',
  '13': 'DT_FINI',
  '14': 'DT_SONAME',
  '15': 'DT_RPATH',
  '16': 'DT_SYMBOLIC',
  '17': 'DT_REL',
  '18': 'DT_RELSZ',
  '19': 'DT_RELENT',
  '20': 'DT_PLTREL',
  '21': 'DT_DEBUG',
  '22': 'DT_TEXTREL',
  '23': 'DT_JMPREL',
  '24': 'DT_BIND_NOW',
  '25': 'DT_INIT_ARRAY',
  '26': 'DT_FINI_ARRAY',
  '27': 'DT_INIT_ARRAYSZ',
  '28': 'DT_FINI_ARRAYSZ',
  '30': 'DT_FLAGS',
  0x6ffffef5: 'DT_GNU_HASH',
  0x6ffffffb: 'DT_FLAGS_1',
  0x6ffffffe: 'DT_VERNEED',
  0x6fffffff: 'DT_VERNEEDNUM',
  0x6ffffff0: 'DT_VERSYM',
  0x6ffffff9: 'DT_RELACOUNT',
  
  DT_NULL: 0,
  DT_NEEDED: 1,
  DT_PLTRELSZ: 2,
  DT_PLTGOT: 3,
  DT_HASH: 4,
  DT_STRTAB: 5,
  DT_SYMTAB: 6,
  DT_RELA: 7,
  DT_RELASZ: 8,
  DT_RELAENT: 9,
  DT_STRSZ: 10,
  DT_SYMENT: 11,
  DT_INIT: 12,
  DT_FINI: 13,
  DT_SONAME: 14,
  DT_RPATH: 15,
  DT_SYMBOLIC: 16,
  DT_REL: 17,
  DT_RELSZ: 18,
  DT_RELENT: 19,
  DT_PLTREL: 20,
  DT_DEBUG: 21,
  DT_TEXTREL: 22,
  DT_JMPREL: 23,
  DT_BIND_NOW: 24,
  DT_INIT_ARRAY: 25,
  DT_FINI_ARRAY: 26,
  DT_INIT_ARRAYSZ: 27,
  DT_FINI_ARRAYSZ: 28,
  DT_FLAGS: 30,
  DT_GNU_HASH: 0x6ffffef5,
  DT_FLAGS_1: 0x6ffffffb,
  DT_VERNEED: 0x6ffffffe,
  DT_VERNEEDNUM: 0x6fffffff,
  DT_VERSYM: 0x6ffffff0,
  DT_RELACOUNT: 0x6ffffff9,
};

module.exports = constants;
